aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java

```diff
@@ -104,8 +112,23 @@ private S3Client client() {
   @Override
   public void initialize(Map<String, String> properties) {
     this.awsProperties = new AwsProperties(properties);
-    this.awsClientFactory = AwsClientFactories.from(properties);
-    this.s3 = awsClientFactory::s3;
+
+    // Do not override s3 client if it was provided
+    if (s3 == null) {
+      this.s3 = AwsClientFactories.from(properties)::s3;
+    }
+
+    // Report Hadoop metrics if Hadoop is available
+    try {
+      Class<? extends MetricsContext> clazz = DynClasses.builder().impl(DEFAULT_METRICS_IMPL).buildChecked();
+
+      this.metrics = clazz.getDeclaredConstructor().newInstance();
+      metrics.initialize(ImmutableMap.of("fileio.scheme", "s3"));
+    } catch (ClassNotFoundException e) {
```

Comment:
I'd recommend using `DynConstructors` rather than `DynClasses` because `DynConstructors` will catch and handle `NoClassDefFoundError` as well as `ClassNotFoundException` and will always throw `ClassNotFoundException`. That handles cases where dependencies of the class aren't found, like when Hadoop is missing.

Otherwise, this may leak the `NoClassDefFoundError` and fail when Hadoop is missing.

I'd also update the error message here to `Unable to load metrics class`. And since this is a warning, consider suppressing the exception because we expect it. Exception traces that are expected tend to confuse people looking at logs.